src/commands/base-command.ts

```diff
@@ -1018,4 +1019,60 @@ export class CommandHelper {
     finished.add(node);
     result.unshift(node);
   }
+
+  public async applyDatacenter(
+    name: string,
+    datacenter: Datacenter,
+    pipeline: Pipeline,
+    logger: winston.Logger | undefined,
+  ): Promise<void> {
+    return pipeline
+      .apply({
+        providerStore: this.providerStore,
+        logger: logger,
+      })
+      .then(async () => {
+        await this.saveDatacenter(name, datacenter, pipeline);
+      })
+      .catch(async (err) => {
+        await this.saveDatacenter(name, datacenter, pipeline);
+        console.error(err);
+        Deno.exit(1);
+      });
+  }
+
+  public async applyEnvironment(
+    name: string,
+    datacenterRecord: DatacenterRecord,
+    environmentRecord: EnvironmentRecord,
+    environment: Environment,
+    pipeline: Pipeline,
+    logger: winston.Logger | undefined,
+  ): Promise<void> {
+    return pipeline
+      .apply({
+        providerStore: this.providerStore,
+        logger,
+      })
+      .then(async () => {
+        await this.saveEnvironment(
+          environmentRecord!.datacenter,
+          name,
+          datacenterRecord.config,
+          environment!,
+          pipeline,
+        );
+      })
+      .catch(async (err) => {
+        await this.saveEnvironment(
+          environmentRecord!.datacenter,
+          name,
+          datacenterRecord.config,
+          environment!,
+          pipeline,
+        );
+        console.error(err);
+        Deno.exit(1);
+      });
```

Comment:
This is a good use case for [`finally()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally)

Reply:
Finally does not work if the catch has `Deno.exit(1);`